lib/datadog/search-syntax/src/vrl.rs

```diff
@@ -0,0 +1,592 @@
+        // If there's one expression remaining, use it as the RHS; no need to wrap.
+        1 => make_op(
+            make_node(expr),
+            op,
+            make_node(exprs.next().expect("must contain expression").into()),
+        ),
+        // For 2+ expressions, recurse over the RHS.
+        _ => make_op(
+            make_node(expr),
+            op,
+            make_node(make_container_group(nest_exprs(exprs, op))),
+        ),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::make_node;
+    use crate::parse;
+    use vrl_parser::ast;
+
+    // Datadog search syntax -> VRL
+    static TESTS: &[(&str, &str)] = &[
+        // Match everything (empty).
+        ("", "exists(.message) || (exists(.custom.error.message) || (exists(.custom.error.stack) || (exists(.custom.title) || exists(._default_))))"),
+        // Match everything.
+        ("*:*", "exists(.message) || (exists(.custom.error.message) || (exists(.custom.error.stack) || (exists(.custom.title) || exists(._default_))))"),
+        // Match everything (negate).
+        ("NOT(*:*)", "!exists(.message) || (!exists(.custom.error.message) || (!exists(.custom.error.stack) || (!exists(.custom.title) || !exists(._default_))))"),
+        // Match nothing.
+        ("-*:*", "!exists(.message) || (!exists(.custom.error.message) || (!exists(.custom.error.stack) || (!exists(.custom.title) || !exists(._default_))))"),
+        // Tag exists.
+        ("_exists_:a", "exists(.a)"),
+        // Tag exists (negate).
+        ("NOT _exists_:a", "!exists(.a)"),
+        // Tag exists (negate w/-).
+        ("-_exists_:a", "!exists(.a)"),
+        // Facet exists.
+        ("_exists_:@b", "exists(.custom.b)"),
+        // Facet exists (negate).
+        ("NOT _exists_:@b", "!exists(.custom.b)"),
+        // Facet exists (negate w/-).
+        ("-_exists_:@b", "!exists(.custom.b)"),
+        // Tag doesn't exist.
+        ("_missing_:a", "!exists(.a)"),
+        // Tag doesn't exist (negate).
+        ("NOT _missing_:a", "!!exists(.a)"),
+        // Tag doesn't exist (negate w/-).
+        ("-_missing_:a", "!!exists(.a)"),
+        // Facet doesn't exist.
+        ("_missing_:@b", "!exists(.custom.b)"),
+        // Facet doesn't exist (negate).
+        ("NOT _missing_:@b", "!!exists(.custom.b)"),
+        // Facet doesn't exist (negate w/-).
+        ("-_missing_:@b", "!!exists(.custom.b)"),
+        // Keyword.
+        ("bla", r#"match(.message, r'\bbla\b')"#),
+        // Keyword (negate).
+        ("NOT bla", r#"!match(.message, r'\bbla\b')"#),
+        // Keyword (negate w/-).
+        ("-bla", r#"!match(.message, r'\bbla\b')"#),
+        // Quoted keyword.
+        (r#""bla""#, r#"match(.message, r'\bbla\b')"#),
+        // Quoted keyword (negate).
+        (r#"NOT "bla""#, r#"!match(.message, r'\bbla\b')"#),
+        // Quoted keyword (negate w/-).
+        (r#"-"bla""#, r#"!match(.message, r'\bbla\b')"#),
+        // Tag match.
+        ("a:bla", r#"match(.a, r'\bbla\b')"#),
```

Comment:
We support tokenised field search(using `\b`) only for "default" fields `message`, `custom.error.message`,  `custom.error.stack` and `custom.title`.  All of them are checked also if no field specified.

Otherwise it should be a full match: a:bla => a == "bla", a:bla* => match(a, "^bla.*$")

Reply:
Ah, that's very interesting.

In that case, I guess it'd be useful to return an enum from `normalize_tags` to distinguish between fields that require word boundary matching and those that don't.

For 'full' matches, we could then get away with `field == "value"` rather than calling out `match` (as your example shows).

I'll update that once the 'default fields' refactor is done.

Reply:
For completion, noting here that is done and exampled in https://github.com/timberio/vector/pull/7633#issuecomment-853925594